iree/hal/local/task_device.c

```diff
@@ -0,0 +1,331 @@
+}
+
+void iree_hal_task_device_params_initialize(
+    iree_hal_task_device_params_t* out_params) {
+  out_params->arena_block_size = 32 * 1024;
+  out_params->queue_count = 8;
+}
+
+static iree_status_t iree_hal_task_device_check_params(
+    const iree_hal_task_device_params_t* params) {
+  if (params->arena_block_size < 512) {
+    return iree_make_status(IREE_STATUS_INVALID_ARGUMENT,
+                            "arena block size too small (< 512 bytes)");
+  }
+  if (params->queue_count == 0) {
+    return iree_make_status(IREE_STATUS_INVALID_ARGUMENT,
+                            "at least one queue is required");
+  }
+  return iree_ok_status();
+}
+
+iree_status_t iree_hal_task_device_create(
+    iree_string_view_t identifier, const iree_hal_task_device_params_t* params,
+    iree_task_executor_t* executor, iree_host_size_t loader_count,
+    iree_hal_executable_loader_t** loaders, iree_allocator_t host_allocator,
+    iree_hal_device_t** out_device) {
+  IREE_ASSERT_ARGUMENT(params);
+  IREE_ASSERT_ARGUMENT(!loader_count || loaders);
+  IREE_ASSERT_ARGUMENT(out_device);
+  *out_device = NULL;
+  IREE_TRACE_ZONE_BEGIN(z0);
+
+  IREE_RETURN_AND_END_ZONE_IF_ERROR(z0,
+                                    iree_hal_task_device_check_params(params));
+
+  iree_hal_task_device_t* device = NULL;
+  iree_host_size_t total_size =
+      sizeof(*device) + params->queue_count * sizeof(*device->queues) +
+      identifier.size + loader_count * sizeof(*device->loaders);
+  iree_status_t status =
+      iree_allocator_malloc(host_allocator, total_size, (void**)&device);
+  if (iree_status_is_ok(status)) {
+    memset(device, 0, total_size);
+    iree_hal_resource_initialize(&iree_hal_task_device_vtable,
+                                 &device->base.resource);
+    iree_string_view_append_to_buffer(
+        identifier, &device->identifier,
+        (char*)device + sizeof(*device) +
+            params->queue_count * sizeof(*device->queues));
+    device->base.host_allocator = host_allocator;
+    iree_arena_block_pool_initialize(1024, host_allocator,
+                                     &device->small_block_pool);
+    iree_arena_block_pool_initialize(params->arena_block_size, host_allocator,
+                                     &device->large_block_pool);
+    device->event_pool = NULL;
+
+    device->executor = executor;
+    iree_task_executor_retain(device->executor);
+
+    device->loader_count = loader_count;
+    device->loaders =
+        (iree_hal_executable_loader_t**)((uint8_t*)device->identifier.data +
+                                         identifier.size);
+    for (iree_host_size_t i = 0; i < device->loader_count; ++i) {
+      device->loaders[i] = loaders[i];
+      iree_hal_executable_loader_retain(device->loaders[i]);
+    }
+
+    device->queue_count = params->queue_count;
+    for (iree_host_size_t i = 0; i < device->queue_count; ++i) {
+      // TODO(benvanik): add a number to each queue ID.
```

Comment:
Not necessarily related to this line, but I would like for more metadata bout devices to be exposed through the API so applications can query device properties. This came up when porting the CTS and I wanted to log some info about each device.

Before (ideally there would be more than a debug string, but that at least exposed a place to put some comments): https://github.com/google/iree/blob/e7df734c536b8cf6c4f8d6b11c69d50d75fb8c2f/iree/hal/cts/driver_test.cc#L33-L36

After (name doesn't indicate much):
https://github.com/google/iree/blob/77fc75dcc41ae637feca57329f96051558ccd0b1/iree/hal/cts/driver_test.cc#L32-L44

Reply:
I intentionally omitted programmatic queries until I get around to exposing these to the compiler so that we can ensure they match. As for debug strings that's just TBD.